File: ICPC/Elimination-Japan/2010/f/F.java
import java.util.*;

class Node {
  ProblemInfo problemInfo;
  String documentCandidate;
  ArrayList<Node> children;
  int depth;
  
  Node(ProblemInfo problemInfo, String documentCandidate, int depth) {
    this.problemInfo = problemInfo;
    this.documentCandidate = documentCandidate;
    this.depth = depth;
    children = null;
  }
  
  public String getCandidate() {
    return documentCandidate;
  }
  
  private static String reverseString(String str) {
    return new StringBuffer(str).reverse().toString();
  }
  
  public String getReverseCandidate() {
    return reverseString(documentCandidate);
  }
  
  public int length() {
    return documentCandidate.length();
  }
  
  public int getDepth() {
    return depth;
  }
  
  public ArrayList<Node> getChildren() {
    if (children == null) {
      children = new ArrayList<Node>();
      for (String piece: problemInfo.getPieces()) {
        for (String merged: Utils.mergeString(documentCandidate, piece, depth*problemInfo.pieceLength()+1, Integer.MAX_VALUE)) {
          if (!problemInfo.containsCandidate(merged) && isLcsUnderX(merged, problemInfo.getDocument(), problemInfo.getD())) {
            problemInfo.addCandidate(merged);
            children.add(new Node(problemInfo, merged, depth+1));
          }
        }
      }
    }
    return children;
  }
  
  private static boolean isLcsUnderX(String candidates, String document, int limit) {
    int[][] table = new int[candidates.length()+1][2*limit+1];
    
    for (int i = -limit; i <= limit; i++) {
      table[0][limit+i] = Math.abs(i);
    }
    
    int minCost = 100;
    for (int i = 0; i < candidates.length(); i++) {
      int matchCost;
      
      matchCost = candidates.charAt(i) != Utils.myCharAt(document, i-limit) ? 1 : 0;
      table[i+1][0] = Math.min(table[i][0]+matchCost, table[i][1]+1);
      minCost = table[i+1][0];
      
      for (int j = -limit+1; j <= limit-1; j++) {
        matchCost = candidates.charAt(i) != Utils.myCharAt(document, i+j) ? 1 : 0;
        table[i+1][limit+j] = Math.min(table[i][limit+j]+matchCost, Math.min(table[i][limit+j+1]+1, table[i+1][limit+j-1]+1));
        minCost = Math.min(table[i+1][limit+j], minCost);
      }
      
      matchCost = candidates.charAt(i) != Utils.myCharAt(document, i+limit) ? 1 : 0;
      table[i+1][2*limit] = Math.min(table[i][2*limit]+matchCost, table[i+1][2*limit-1]+1);
      minCost = Math.min(table[i+1][2*limit], minCost);
      
      if (minCost > limit) {
        return false;
      }
    }
    return minCost <= limit;
  }
}

class ProblemInfo {
  String document;
  String[] pieces;
  int d;
  HashSet<String> candidates;
  
  ProblemInfo(String document, String[] pieces, int d) {
    this.document = document;
    this.pieces = pieces;
    this.d = d;
    candidates = new HashSet<String>();
  }
  
  private static String reverseString(String str) {
    return new StringBuffer(str).reverse().toString();
  }
  
  private String[] reversePieces() {
    String[] output = new String[pieces.length];
    for (int i = 0; i < pieces.length; i++) {
      output[i] = reverseString(pieces[i]);
    }
    return output;
  }
  
  public String[] getPieces() {
    return pieces;
  }
  
  public ProblemInfo makeReverseProblemInfo() {
    return new ProblemInfo(reverseString(document), reversePieces(), d);
  }
  
  public String getDocument() {
    return document;
  }
  
  public int getD() {
    return d;
  }
  
  public int pieceLength() {
    return pieces[0].length();
  }
  
  public boolean containsCandidate(String candidate) {
    return candidates.contains(candidate);
  }
  
  public void addCandidate(String candidate) {
    candidates.add(candidate);
  }
}

class Utils {
  public static char myCharAt(String s, int index) {
    return index < 0 || s.length() <= index ? '-' : s.charAt(index);
  }
  
  public static ArrayList<String> mergeString(String head, String tail, int minLength, int maxLength) {
    int minMergedLength = Math.max(head.length(), tail.length());
    int minTargetLength = Math.max(minLength, minMergedLength);
    int maxMergedLength = head.length() + tail.length();
    int maxTargetLength = Math.min(maxMergedLength, maxLength);
    int minCommonLength = maxMergedLength - maxTargetLength;
    int maxCommonLength = maxMergedLength - minTargetLength;
    
    ArrayList<String> output = new ArrayList<String>();
    for (int commonLength = minCommonLength; commonLength <= maxCommonLength; commonLength++) {
      if (head.endsWith(tail.substring(0, commonLength))) {
        output.add(head + tail.substring(commonLength));
      }
    }
    return output;
  }
}

public class F {
  private static boolean isLcsUnderX(String candidate, String document, int limit) {
    if (Math.abs(candidate.length() - document.length()) > limit) {
      return false;
    }
    
    int[][] table = new int[candidate.length()+1][2*limit+1];
    for (int i = -limit; i <= limit; i++) {
      table[0][limit+i] = Math.abs(i);
    }

    int minCost = 100;
    for (int i = 0; i < candidate.length(); i++) {
      int matchCost;
      
      matchCost = candidate.charAt(i) != Utils.myCharAt(document, i-limit) ? 1 : 0;
      table[i+1][0] = Math.min(table[i][0]+matchCost, table[i][1]+1);
      minCost = table[i+1][0];
      
      for (int j = -limit+1; j <= limit-1; j++) {
        matchCost = candidate.charAt(i) != Utils.myCharAt(document, i+j) ? 1 : 0;
        table[i+1][limit+j] = Math.min(table[i][limit+j]+matchCost, Math.min(table[i][limit+j+1]+1, table[i+1][limit+j-1]+1));
        minCost = Math.min(table[i+1][limit+j], minCost);
      }
      
      matchCost = candidate.charAt(i) != Utils.myCharAt(document, i+limit) ? 1 : 0;
      table[i+1][2*limit] = Math.min(table[i][2*limit]+matchCost, table[i+1][2*limit-1]+1);
      minCost = Math.min(table[i+1][2*limit], minCost);
      
      if (minCost > limit) {
        return false;
      }
    }
    return table[candidate.length()][limit+document.length()-candidate.length()] <= limit;
  }
  
  private static void calculateDocumentCandidates(ProblemInfo problemInfo, Node forward, Node backward, HashSet<String> candidates) {
    if (forward.length() + backward.length() >= problemInfo.getDocument().length()-2) {
      ArrayList<String> mergedList = Utils.mergeString(forward.getCandidate(), backward.getReverseCandidate(), problemInfo.getDocument().length()-2, problemInfo.getDocument().length()+2);
      for (String merged: mergedList) {
        if (!candidates.contains(merged) && isLcsUnderX(merged, problemInfo.getDocument(), problemInfo.getD())) {
          candidates.add(merged);
        }
      }
      if (forward.length() + backward.length() >= problemInfo.getDocument().length()+2) {
        return;
      }
    }
    
    if (forward.getDepth() == backward.getDepth()) {
      for (Node child: forward.getChildren()) {
        calculateDocumentCandidates(problemInfo, child, backward, candidates);
      }
    } else {
      for (Node child: backward.getChildren()) {
        calculateDocumentCandidates(problemInfo, forward, child, candidates);
      }
    }
  }
  
  private static void printAnswers(HashSet<String> answerSet) {
    System.out.print(answerSet.size() + "\n");
    if (answerSet.size() <= 5) {
      String[] array = new String[answerSet.size()];
      array = answerSet.toArray(array);
      Arrays.sort(array);
      for (String str: array) {
        System.out.print(str + "\n");
      }
    }
  }
  
  private static void solve(ProblemInfo forwardInfo) {
    HashSet<String> answerSet = new HashSet<String>();
    calculateDocumentCandidates(forwardInfo, new Node(forwardInfo, "", 0), new Node(forwardInfo.makeReverseProblemInfo(), "", 0), answerSet);
    printAnswers(answerSet);
  }
  
  public static void main(String[] args) {
    Scanner s = new Scanner(System.in);
    while (true) {
      
      int d = s.nextInt();
      int n = s.nextInt();
      if (d == 0) {
        break;
      }
      String document = s.next();
      String[] pieces = new String[n];
      for (int i = 0; i < n; i++) {
        pieces[i] = s.next();
      }
      solve(new ProblemInfo(document, pieces, d));
    }
  }
}

File: ICPC/Elimination-Japan/2010/f/f.cpp
#include<cstdlib>
#include<iostream>
#include<set>
#include<vector>

using namespace std;

bool mergable(int length, const string &head, const string &tail) {
  for (int i = 0; i < length; i++) {
    if (head[head.length()-length+i] != tail[i]) return false;
  }
  return true;
}

vector<string> getNextSuffixes(const string &head, const string &tail) {
  vector<string> output;
  for (int i = 0; i < tail.length(); i++) {
    if (mergable(i, head, tail)) output.push_back(tail.substr(i));
  }
  return output;
}

char myCharAt(const string &document, int index) {
  return 0 <= index && index < document.length() ? document[index] : '-';
}

pair<bool, vector<int> > calculateLcsTable(int previousLength, const string &suffix, const vector<int> &previousLcsTable, const string &document, int d) {
  vector<int> table1 = previousLcsTable;
  for (int i = 0; i < suffix.length(); i++) {
    vector<int> table2(2*d+1);
    int minCost = 10;
    for (int j = -d; j <= d; j++) {
      table2[d+j] = table1[d+j] + (suffix[i] != myCharAt(document, previousLength+i+j) ? 1 : 0);
      if (d+j-1 >= 0) table2[d+j] = min(table2[d+j], table2[d+j-1]+1);
      if (d+j+1 < 2*d+1) table2[d+j] = min(table2[d+j], table1[d+j+1]+1);
      minCost = min(table2[d+j], minCost);
    }
    
    if (minCost > d) return make_pair(false, vector<int>());
    table1 = table2;
  }
  return make_pair(true, table1);
}

void dfs(const string &piece, const vector<int> &previousLcsTable, set<string> &visited, set<string> &answer, int d, int n, const string &document, const string* pieces) {
  if (piece.length() > document.length()+2) return;
  visited.insert(piece);
  if (piece.length() >= document.length()-2) {
    if (previousLcsTable[d+document.length()-piece.length()] <= d) answer.insert(piece);
  }
  
  for (int i = 0; i < n; i++) {
    vector<string> nexts = getNextSuffixes(piece, pieces[i]);
    for (vector<string>::iterator it = nexts.begin(); it != nexts.end(); it++) {
      string nextPiece = piece + *it;
      if (visited.find(nextPiece) == visited.end()) {
        pair<bool, vector<int> > p = calculateLcsTable(piece.length(), *it, previousLcsTable, document, d);
        if (p.first) {
          dfs(nextPiece, p.second, visited, answer, d, n, document, pieces);
        }
      }
    }
  }
}

void solve(int d, int n, const string &document, const string* pieces) {
  vector<int> initialLcsTable;
  for (int i = -d; i <= d; i++) initialLcsTable.push_back(abs(i));
  
  set<string> answers;
  set<string> visited;
  dfs("", initialLcsTable, visited, answers, d, n, document, pieces);
  cout << answers.size() << endl;
  if (answers.size() > 5) return;
  for (set<string>::iterator it = answers.begin(); it != answers.end(); it++) cout << *it << endl;
}

int main(void) {
  while (true) {
    int d, n;
    cin >> d >> n;
    if (d == 0) break;
    string document;
    cin >> document;
    string pieces[n];
    for (int i = 0; i < n; i++) cin >> pieces[i];
    solve(d, n, document, pieces);
  }
}
